ci: Invalidate depends caches when sources have been changed

hebasto opened the pull request:
On master (502d22ceed1f90ed41336260f8eb428d3acaf514) the Cirrus CI do _not_ invalidate depends caches when their sources has been changed, i.e. source version updates, a new dependency etc.

This behavior causes:
- breaking CI build for Android APK  (see #22708)
- the cache sizes growing

It is assumed that caches could be invalidated via Cirrus CI API/GUI, but it is inconvenient and it does not work as expected in all cases.

The common part of `fingerprint_key`s for both `depends_sources_cache` and `depends_built_cache` is the recent commit that changes sources in the `src/depends` sub-directory. 

For `depends_built_cache` additionally `CIRRUS_TASK_NAME` is used to avoid sharing of the same cache between tasks with different OSes and different `DEP_OPTS`.

---

The `depends_sources` cache:
- in master (502d22ceed1f90ed41336260f8eb428d3acaf514)

![Screenshot from 2021-08-16 11-10-17](https://user-images.githubusercontent.com/32963518/129532207-56c09e06-c8d0-4f1e-a692-f198da011b20.png)

- with this PR

![Screenshot from 2021-08-16 11-10-42](https://user-images.githubusercontent.com/32963518/129532253-f5426814-4f58-448f-ad52-da26cc312af7.png)

MarcoFalke commented on .cirrus.yml:
```diff
@@ -213,6 +215,7 @@ task:
   << : *DEPENDS_SDK_CACHE_TEMPLATE
   depends_sources_cache:
     folder: "depends/sources"
+    fingerprint_script: echo $SOURCES_CACHE_FINGERPRINT_KEY
```
why not `git log -1 ./depends`?

hebasto replied:
Should we invoke it within a Docker image, i.e., with `DOCKER_EXEC`?

MarcoFalke replied:
It should work as long as `BASE_TEMPLATE` is run before `git` is called

hebasto replied:
Thanks! Done in the recent push.

hebasto replied:
@MarcoFalke 
> why not `git log -1 ./depends`?

It works fine for `depends_sources_cache`, but not for `depends_built_cache` as the the actual content of the latter depends on the `DEP_OPTS` variable.

hebasto replied:
> It works fine for `depends_sources_cache`, but not for `depends_built_cache` as the the actual content of the latter depends on the `DEP_OPTS` variable.

Addressed in the recent push. OP has been updated.

hebasto committed:
ci: Reorder scripts to make git available before depends_sources_cache

hebasto committed:
ci: Invalidate depends caches when sources have been changed

jonatack commented:
Concept ACK

Zero-1729 commented:
Approach ACK

hebasto commented:
CI is green, and this PR is ready for reviewing :)

laanwj commented on .cirrus.yml:
```diff
@@ -41,9 +40,14 @@ global_task_template: &GLOBAL_TASK_TEMPLATE
     folder: "/tmp/ccache_dir"
   depends_built_cache:
     folder: "depends/built"
+    fingerprint_script: echo $CIRRUS_TASK_NAME $(git log -1 --oneline ./depends)
```
Might use `git rev-list -1 HEAD ./depends` here; this prints only the (full, independent of `core.abbrev`) commit hash.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/22710#issuecomment-901260528).

hebasto commented:
Updated 93f3427a47bf1761e4c66f9f6c5a0c09728e3a0e -> f52a72af569526f319d9f1480eb473b9570125c3 ([pr22710.04](https://github.com/hebasto/bitcoin/commits/pr22710.04) -> [pr22710.05](https://github.com/hebasto/bitcoin/commits/pr22710.05), [diff](https://github.com/hebasto/bitcoin/compare/pr22710.04..pr22710.05)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/22710#discussion_r691209537)

MarcoFalke commented:
cr ACK f52a72af569526f319d9f1480eb473b9570125c3

Zero-1729 reviewed (commented):
crACK f52a72af569526f319d9f1480eb473b9570125c3
